### ISSUE TITLE
UnknownFields in the serializer seems to be unused, can this be removed or is it unimplemented?

### ISSUE THREAD

**Issue by 7sharp9** [ctaggart/froto#84]

**Comment by jhugard** [comment]
Not using the UnknownFields list in the Serializer tests was an oversight.  The intention with this list is to allow pass-thru of fields that were added after the code was written, in order to survive protocol evolution.

So, the deserializer will record such fields on UnknownFields, then the generated serializer should be written to reintroduce those fields to the binary output, preserving the complete and possibly modified message (if not original order, which IIRC is OK per the spec).

E.g., [these lines](https://github.com/ctaggart/froto/blob/ba4d5e506e1be6105dc6746a08a1d6714793e8fe/Serialization.Test/TestSerialization.fs#L28-L31) should probably be written as:

```F#
            static member Serializer (m, zcb) =
                (m.id            |> Encode.fromVarint 1) >>
                (m.name          |> Encode.fromString 2) >>
                (m._unknownFields|> Encode.fromRawFields)
                <| zcb
```

Note the introduction of `(m._unknownFields|> Encode.fromRawFields)`.

**Comment by 7sharp9** [comment]
So all the unknowns are at the end of the stream, ok.  

Im not sure when yet, but I could add this to the records serialization test and serialization.fs file at some point.

**Comment by 7sharp9** [comment]
Its a pity that found and unknown fields have to be part of the record definition it makes the construction of the port record a little more tedious:

e.g. It would be nice to not have `_unknownFields` and `_foundFields` I suppose theres no way around this with the record implementation in F#.

```
let bundle = { martId = 42
                 member_id = "memberId"
                 channel_type = "myChannel"
                 skus =  ["retailsku1"; "retailsku2"]
                 _unknownFields = []
                 _foundFields = Set.empty }
```

**Comment by jhugard** [comment]
Cool.  See the [proto3](https://developers.google.com/protocol-buffers/docs/proto3#unknowns) and [proto2](https://developers.google.com/protocol-buffers/docs/proto#updating) language guides for notes about handling unknown fields.

Also note that while the example Serializer code does preserve order of repeated fields, it does not preserve order between different fields: it always writes in the order defined in the Serializer function (generally ascending field number order).  Pretty sure this is consistent with other protobuf implementations.

**Comment by jhugard** [comment]
You could also use Default, which allows the record to be expanded later without necessitating code changes (the new fields would use defaults).  But then the compiler won't point out these cases, either:

```F#
let bundle = { Bundle.Default with
                 martId = 42
                 member_id = "memberId"
                 channel_type = "myChannel"
                 skus =  ["retailsku1"; "retailsku2"] }
```

**Comment by jhugard** [comment]
Well, if your codegen had an option to generate `proto3` specific serdes, then you _could_ drop both: preserving unknown fields isn't mandatory for proto3, and proto3 does not support the `required` keyword, so `_foundFields` isn't needed.

**Comment by jhugard** [comment]
Update: the ExampleProtoRecord.fs _does_ include serialization of unknown fields.  See [here](https://github.com/ctaggart/froto/blob/bd78b5aa4b0f3dee51199dcdfe221eb7daa622b1/Serialization.Test/ExampleProtoRecord.fs#L43) (I thought I had an example of that).

**Comment by 7sharp9** [comment]
Yeah I did write the copy and update method with the record, it just feels a bit clunky thats all.  

Im only really interested in the proto3 version to be honest, I think one of the C# implementations dropped all the proto2 parts for simplicity.  

Im trying to thinks of a way of removing the SRTP constraints so that the RememberFound, FoundFields and UnknownFields are no longer required in the record definition.

**Comment by 7sharp9** [comment]
I wonder if I could add a proto3 specific module for the serialization of proto3 only, i.e. remove the constraints.  Alternatively the constrains could just be proto2/proto3 interfaces that could be used rather then SRTP constraints.

That way I could generate serialize/deserialize  that were more compact and made just for proto3.  Im just thinking out loud really as I can do all this in my form but then it won't allow me to contribute back here.

**Comment by jhugard** [comment]
Proto3 files should/must have `syntax=proto3;` near the start of the file and I'm certain I record this as part of the AST.  You could gate codegen based on this to either provide UnknownFields (and Required Fields) support, or not, based on that flag.

**Comment by 7sharp9** [comment]
ok, let me put something together.

**Comment by 7sharp9** [comment]
Technically proto2 messages can be constructed with no UnknownFields as there is no constraint to require this, its entirely up to the implementor or codegen.

**Comment by 7sharp9** [comment]
So the serializer method would have to change to something like this  to enforce it:
```
    /// Serialize message into a ZeroCopyBuffer.
    let inline toZeroCopyBuffer m zcb =
        (^msg : (static member Serializer : ^msg * ZeroCopyBuffer -> ZeroCopyBuffer) (m, zcb))
        let unknownFields = (^msg : (static member UnknownFields: ^msg -> List<RawField>) (m) )
        Encode.fromRawFields unknownFields zcb 
```

**Comment by jhugard** [comment]
Fixed in commit ca65a0f6974b4ee9e4f0c143e620428f6b9e7e82
